agent/pkg/singularity/singularity.go

```diff
@@ -0,0 +1,696 @@
+package singularity
+
+import (
+	"bufio"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/url"
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	dcontainer "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/determined-ai/determined/agent/internal/container"
+	"github.com/determined-ai/determined/agent/internal/options"
+	"github.com/determined-ai/determined/agent/pkg/docker"
+	"github.com/determined-ai/determined/agent/pkg/events"
+	"github.com/determined-ai/determined/master/pkg/aproto"
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/master/pkg/cproto"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/syncx/waitgroupx"
+	"github.com/determined-ai/determined/master/pkg/tasks"
+)
+
+var singularityWrapperEntrypoint = path.Join(tasks.RunDir, tasks.SingularityEntrypointWrapperScript)
+
+const (
+	// TODO(DET-9111): Parameterize this by agent ID.
+	agentTmp         = "/tmp/determined/agent"
+	hostNetworking   = "host"
+	bridgeNetworking = "bridge"
+	envFileName      = "envfile"
+	archivesName     = "archives"
+)
+
+// SingularityContainer captures the state of a container.
+type SingularityContainer struct {
+	PID     int            `json:"pid"`
+	Req     cproto.RunSpec `json:"req"`
+	TmpDir  string         `json:"tmp_dir"`
+	Proc    *os.Process    `json:"-"`
+	Started atomic.Bool    `json:"started"`
+}
+
+// SingularityClient implements ContainerRuntime.
+type SingularityClient struct {
+	log        *logrus.Entry
+	opts       options.SingularityOptions
+	mu         sync.Mutex
+	wg         waitgroupx.Group
+	containers map[cproto.ID]*SingularityContainer
+}
+
+// New returns a new singularity client, which launches and tracks containers.
+func New(opts options.SingularityOptions) (*SingularityClient, error) {
+	if err := os.RemoveAll(agentTmp); err != nil {
+		return nil, fmt.Errorf("removing agent tmp from previous runs: %w", err)
+	}
+
+	if err := os.MkdirAll(agentTmp, 0o700); err != nil {
+		return nil, fmt.Errorf("preparing agent tmp: %w", err)
+	}
+
+	return &SingularityClient{
+		log:        logrus.WithField("compotent", "singularity"),
```

Comment:
compotent?

Reply:
ah.. yes. i will rebase and fix this, then land.